src/config/storage/json_storage.rs

```diff
@@ -0,0 +1,102 @@
+use crate::config::settings::Setting;
+use crate::config::StorageAdapter;
+use std::collections::HashMap;
+use std::fs;
+use std::fs::File;
+use std::io::{Read, Seek, Write};
+use std::sync::{Arc, Mutex};
+use serde_json::Value;
+
+pub struct JsonStorageAdapter {
+    path: String,
+    file_mutex: Arc<Mutex<File>>,
+    elements: HashMap<String, Setting>
+}
+
+impl JsonStorageAdapter {
+    pub fn new(path: &str) -> Self {
+        let file = match fs::metadata(path) {
+            Ok(metadata) => {
+                if ! metadata.is_file() {
+                    panic!("json file is not a regular file")
+                }
+
+                File::options()
+                    .read(true)
+                    .write(true)
+                    .open(path)
+                    .expect("failed to open json file")
+            }
+            Err(_) => {
+                let json = "{}";
+
+                let mut file = File::create(path).expect("cannot create json file");
+                file.write_all(json.as_bytes()).expect("failed to write to file");
+
+                file
+            }
+        };
+
+        let mut adapter = JsonStorageAdapter {
+            path: path.to_string(),
+            file_mutex: Arc::new(Mutex::new(file)),
+            elements: HashMap::new(),
+        };
+
+        adapter.read_file();
+
+        adapter
+    }
+}
+
+impl StorageAdapter for JsonStorageAdapter {
+    fn get_setting(&self, key: &str) -> Option<Setting> {
+        match self.elements.get(key) {
```

Comment:
basically the json adapter reads the whole json during init, and not during every get_setting call.. This would create a secondary layer of caching and doesn't reflect any changes that are done on disk. We might want to reread the whole json on each json file, OR do something smarter like checking access/modify time